File: WinForms/VB/ChartsMainDemo/Modules/Highlighted/RealTimeChart.vb
Imports System
Imports System.Linq

Namespace DevExpress.XtraCharts.Demos

    Public Partial Class RealTimeChartDemo
        Inherits ChartDemoModule

        Private ReadOnly fpsBuffer As Double() = New Double(9) {}

        Private lastIndex As Integer = 0

        Private lastTime As Date = Date.Now

        Private lastFPSUpdateTime As Double = 0

        Private ReadOnly dataGenerator As SensorDataGenerator

        Friend Overrides ReadOnly Property ChartDesignerEnabled As Boolean
            Get
                Return False
            End Get
        End Property

        Friend Overrides ReadOnly Property ChartControl As ChartControl
            Get
                Return chart
            End Get
        End Property

        Friend Overrides ReadOnly Property ExportedObject As Object
            Get
                Return Nothing
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
            ChartControl.BeginInit()
            dataGenerator = New SensorDataGenerator()
            dataGenerator.GenerateInitialData()
            ChartControl.DataSource = dataGenerator.DataSource
            ColorizeSeries()
            dataGenerator.Start()
            synchronousTimer.Enabled = True
            ChartControl.EndInit()
            Dim axisX As SwiftPlotDiagramAxisX = CType(ChartControl.Diagram, SwiftPlotDiagram).AxisX
            axisX.VisualRange.SetMinMaxValues(CDate(axisX.WholeRange.MaxValue) - New TimeSpan(0, 0, 10), (CDate(axisX.WholeRange.MaxValue)).AddSeconds(1))
        End Sub

        Private Sub ColorizeSeries()
            Dim paletteEntries As PaletteEntry() = ChartControl.GetPaletteEntries(2)
            ChartControl.Series(0).View.Color = paletteEntries(0).Color
            ChartControl.Series(1).View.Color = paletteEntries(1).Color
            ChartControl.Series(2).View.Color = paletteEntries(0).Color
            ChartControl.Series(3).View.Color = paletteEntries(1).Color
            ChartControl.Series(4).View.Color = paletteEntries(0).Color
            ChartControl.Series(5).View.Color = paletteEntries(1).Color
            ChartControl.Series(6).View.Color = paletteEntries(0).Color
            ChartControl.Series(7).View.Color = paletteEntries(1).Color
            ChartControl.Legend.CustomItems(0).MarkerColor = paletteEntries(0).Color
            ChartControl.Legend.CustomItems(1).MarkerColor = paletteEntries(1).Color
        End Sub

        Private Sub synchronousTimer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            If dataGenerator IsNot Nothing Then dataGenerator.UpdateDataSource()
        End Sub

        Private Sub Chart_CustomPaint(ByVal sender As Object, ByVal e As CustomPaintEventArgs)
            Dim currentTime As Date = Date.Now
            Dim diff As Double = Math.Max(1, (currentTime - lastTime).TotalMilliseconds)
            lastTime = currentTime
            If diff <> 1 Then fpsBuffer(Math.Min(Threading.Interlocked.Increment(lastIndex), lastIndex - 1)) = 1000 / diff
            If lastFPSUpdateTime = 0 OrElse lastFPSUpdateTime > 100 Then
                chart.Titles(1).Text = String.Format("FPS: {0:0.00}", fpsBuffer.Average())
                lastFPSUpdateTime = 0
            End If

            lastFPSUpdateTime += diff
            If lastIndex >= fpsBuffer.Length Then lastIndex = 0
        End Sub

        Protected Overrides Sub OnPaletteChanged()
            MyBase.OnPaletteChanged()
            ColorizeSeries()
        End Sub

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing AndAlso components IsNot Nothing AndAlso dataGenerator IsNot Nothing Then
                dataGenerator.Stop()
                components.Dispose()
            End If

            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace

File: WinForms/CS/ChartsMainDemo/Modules/Highlighted/RealTimeChart.cs
using System;
using System.Linq;
using DevExpress.DXperience.Demos;

namespace DevExpress.XtraCharts.Demos {
    public partial class RealTimeChartDemo : ChartDemoModule {
        readonly double[] fpsBuffer = new double[10];
        int lastIndex = 0;
        DateTime lastTime = DateTime.Now;
        double lastFPSUpdateTime = 0;
        readonly SensorDataGenerator dataGenerator;

        internal override bool ChartDesignerEnabled { get { return false; } }
        internal override ChartControl ChartControl { get { return chart; } }
        internal override object ExportedObject { get { return null; } }

        public RealTimeChartDemo() {
            InitializeComponent();
            ChartControl.BeginInit();
            dataGenerator = new SensorDataGenerator();
            dataGenerator.GenerateInitialData();
            ChartControl.DataSource = dataGenerator.DataSource;
            ColorizeSeries();
            dataGenerator.Start();
            synchronousTimer.Enabled = true;
            ChartControl.EndInit();
            SwiftPlotDiagramAxisX axisX = ((SwiftPlotDiagram)ChartControl.Diagram).AxisX;
            axisX.VisualRange.SetMinMaxValues((DateTime)axisX.WholeRange.MaxValue - new TimeSpan(0, 0, 10), ((DateTime)axisX.WholeRange.MaxValue).AddSeconds(1));
        }
        void ColorizeSeries() {
            PaletteEntry[] paletteEntries = ChartControl.GetPaletteEntries(2);
            ChartControl.Series[0].View.Color = paletteEntries[0].Color;
            ChartControl.Series[1].View.Color = paletteEntries[1].Color;
            ChartControl.Series[2].View.Color = paletteEntries[0].Color;
            ChartControl.Series[3].View.Color = paletteEntries[1].Color;
            ChartControl.Series[4].View.Color = paletteEntries[0].Color;
            ChartControl.Series[5].View.Color = paletteEntries[1].Color;
            ChartControl.Series[6].View.Color = paletteEntries[0].Color;
            ChartControl.Series[7].View.Color = paletteEntries[1].Color;
            ChartControl.Legend.CustomItems[0].MarkerColor = paletteEntries[0].Color;
            ChartControl.Legend.CustomItems[1].MarkerColor = paletteEntries[1].Color;
        }
        void synchronousTimer_Tick(object sender, EventArgs e) {
            if(dataGenerator != null)
                dataGenerator.UpdateDataSource();
        }
        void Chart_CustomPaint(object sender, CustomPaintEventArgs e) {
            DateTime currentTime = DateTime.Now;
            double diff = Math.Max(1, (currentTime - lastTime).TotalMilliseconds);
            lastTime = currentTime;
            if(diff != 1)
                fpsBuffer[lastIndex++] = 1000 / diff;
            if(lastFPSUpdateTime == 0 || lastFPSUpdateTime > 100) {
                chart.Titles[1].Text = string.Format("FPS: {0:0.00}", fpsBuffer.Average());
                lastFPSUpdateTime = 0;
            }
            lastFPSUpdateTime += diff;
            if(lastIndex >= fpsBuffer.Length)
                lastIndex = 0;
        }
        protected override void OnPaletteChanged() {
            base.OnPaletteChanged();
            ColorizeSeries();
        }
        protected override void Dispose(bool disposing) {
            if(disposing && components != null && dataGenerator != null) {
                dataGenerator.Stop();
                components.Dispose();
            }
            base.Dispose(disposing);
        }
    }
}
